### DISCUSSION TITLE
I want to make abstraction out of Stack.Screen but i can't get the types work properly

### DISCUSSION THREAD

**Discussion by bryanprimus** [react-navigation/react-navigation#11164] [Q&A]
I have this list of screens in an array

```ts
const screens: [NEED_TO_PROVIDE_CORRECT TYPE] = [
  {
    name: 'Home',
    component: HomeScreen,
    // so that i can get intellisense and autocompletion here for Stack.Screen props
    // something like
    // options: {} etc
  },
  {
    name: 'Profile',
    component: ProfileScreen,
  },
  {
    name: 'Settings',
    component: SettingsScreen,
  },
];
```

I want to achieve something like this where I can just spread the screen where it is the props with proper types

```tsx
function App() {
  return (
    <NavigationContainer>
      <Stack.Navigator>
        {screens.map(screen => (
          <Stack.Screen {...screen} />
        ))}
      </Stack.Navigator>
    </NavigationContainer>
  );
}
```

I have checked this https://reactnavigation.org/docs/typescript/ but I still couldn't find a way to achieve my requirement above.

i have tried this, but still doesn't work as i expected.

```ts
type StackParamList =  {
    Home: undefined;
    Profile: undeifned;
    Settings: undefined;
}

type ScreenProps = RouteConfig<
  StackParamList,
  keyof StackParamList,
  NavigationState<StackParamList>,
  {},
  EventMapBase
>;

const screens: ScreenProps = [{name: "[STILL_DIDN'T_GET_THE_TYPE CORRECT_AND_NO_AUTO_COMPLETION]"}];
```

**Comment by bryanprimus** [comment]
i'm able to do it this way


```ts
type ScreenProps<
  T extends keyof RootStackParamList = keyof RootStackParamList,
> = RouteConfig<
  RootStackParamList,
  T,
  StackNavigationState<RootStackParamList>,
  NativeStackNavigationOptions,
  NativeStackNavigationEventMap
>;
```

like this

```ts
type HomeScreenProps = ScreenProps<"Home">
```

but i'm having trouble using it in object since it can't read what data being passed to object

```ts
const screen: RootScreenProps<I_WANT_THIS_TO_READ_THE_VALUE_OF_THE_NAME_KEY_BELOW> = {
  name: 'Home',
  component: CompTest,
  // should have props based on what value provided in the name key
};
```

is there a way to do this in typescript??

The only possible hack i can think of is creating a function from it

like this

```ts
const createScreen = <T extends keyof RootStackParamList>(
  props: RootScreenProps<T>,
): RootScreenProps<T> => {
  return props;
};
```

and using it like this

this will be correctly typed

```
const screen = createScreen({name: 'Home', component: CompTest});
```

but i don't want to use / pass it as a function.

**Comment by satya164** [comment]
> i have tried this, but still doesn't work as i expected.

It didn't work because your `screens` is an array of configs, however `RouteConfig` is the type of the config object.

You can do this instead of using `RouteConfig` type directly:

```js
type StackParamList = {
  Home: undefined;
  Profile: undefined;
  Settings: undefined;
};

const MyStack = createStackNavigator<StackParamList>();

const screens: React.ComponentProps<typeof MyStack['Screen']>[] = [
  {
    name: 'Home',
    component: HomeScreen,
  },
  {
    name: 'Profile',
    component: ProfileScreen,
  },
  {
    name: 'Settings',
    component: SettingsScreen,
  },
];
```

**Reply by bryanprimus** [reply]
it works well with the above StackParamList. but I have an issue with this RootStackParamList

```ts
export type HomeTabParamList = {
  HomeScreen: undefined;
  ProfileScreen: undefined;
};

export type RootStackParamList = {
  HomeTab: NavigatorScreenParams<HomeTabParamList>;
  DetailsScreen: {
    id: string;
  };
};
const RootStack = createNativeStackNavigator<RootStackParamList>();
```

below code will be fine
```ts
const screens: React.ComponentProps<typeof RootStack['Screen']>[] = [
  {
    name: 'HomeTab',
    component: HomeTab,
  },
  {
    name: 'DetailsScreen',
    component: DetailsScreen,
  },
];
```

but other props that rely on `RouteName` types will not work correctly.

this props below will not typed correctly
![image](https://user-images.githubusercontent.com/46083126/213350334-001e984c-9e18-4626-ba7c-e2a5ec312264.png).

in the above `RootStackParamList`
DetailsScreen has params id, and HomeTab has params Nested Navigator Params

for every screen route and navigation will not typed correctly.



![image](https://user-images.githubusercontent.com/46083126/213352315-ea3d49ee-a21e-46cd-93fb-cd283a037549.png)

route type will be `RouteProp<RootStackParamList, keyof RootStackParamList>` where i want it to be `RouteProp<RootStackParamList, "[SCREEN_NAME_SO_THAT_IT_CAN_READ_THE_TYPES_CORRECTLY]">`
navigation type will be `any`

**Reply by 3520kramer** [reply]
@satya164 I am stumpling into the same issue as @bryanltobing. When defining the routes in an array of type `React.ComponentProps<typeof RootStack['Screen']>[]`, then there is issues with the type of the route name when using the navigation and route properties in a screen component. I have added a minimal reproduction of the issue in a [TS playground](https://www.typescriptlang.org/play?jsx=3#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4AoUSWOAbzgDVgkB3OAXzm1wKPRgC0AOxQxgANyRlK4aPHpo+MJAGUY6ANYA5FOOABzUdAA0cNZpWKkSIQAUcYAM4cuOPPgACfDMN0HRwBBCAPSO6mga0uQwAJ5gSIgQEDDmEbYoUCggzgC8dORwhXAAEriqVjYAXHAArkIAJkiYwEJI9RRFZkgwYkL6jpZEVbUNTS1tFOwU0XEJpSDlQ3YOufmdInqGYkHVqRqD1ssQTgA8CEkp4Rrpmdmm+POLh-gAfADa+Bv+20L4ALodIo4GrKXZXA42ezHRxnC57G5ZRz3R4Q37vQgQEFSAHkKbkNBBMIlMqo6rIfgAOgAYgBhE4oipHJwvOB5AAUYBWAEpWSyTkxWHBgi9prF4l0ei1+qioU5WWsil8toEhNUUEIYoDCsDQXB1ZrcRQCUIiSpur1pYyycQYNS6WbJX0BozZY4WezOdCeTk+QK2MLpsaiedknt5VZRKorjpNkYoLDQ1cEdkXmyuYHCfBHIzHNbKTTcJBWkIYK6TmKkBBMIlE5oPqj-u8-vK3gUirQ251CiIFtUHiTGfhjJ2uwTZMWYNUGUth512LP2yP1lkkH2HRbnUsh0uimOizZJxKN6iF4V5+QcfjMzWYDHvtB5WneQrtd0alAhHA2TvCgnLpoKTvLZoBeH8u1obMlkcCkQBQMA2TZMJNHOLFvRZb8u0wrs-z2ClUTgDQkBiHIIKuFDlApHskE4WgKTopCInI6ihVArCuy5Ll2DAk5ghDf8IkAvxgKgViinTQ1yCAA)

Do you have an idea how to solve this? Any help would be greatly appreciated!
